src/networking.c

```diff
@@ -1579,6 +1580,10 @@ int writeToClient(client *c, int handler_installed) {
             return C_ERR;
         }
     }
+
+    if (c->obuf_soft_limit_reached_time && !checkClientOutputBufferLimits(c))
+        queueClientForReprocessing(c);
```

Comment:
ok, i'll create another list for these clients so that we don't mix two different things.
i'm not sure we wanna process just one client per even loop though, i think we wanna be fair to all.
i also think that in case they're really network bound, they won't all get unblocked at the same time (same event loop).
but in case they just got bound by the initial write to the output buffer, they'll all get unblocked as soon as we write to the socket and the OS "swallows" what it can buffer.

Reply:
re: "fair to all"

I know it will be difficult to convince on this point, but processing only 1 in each event loop is "fair", and can result in overall better latencies.  During normal processing, the event loop can spin very fast - so there's really no difference in processing 1 per loop or N per loop - it just changes the spin-rate of the loop.

For these particular clients (who are executing big commands), they are likely to take longer to process than a typical client.  By limiting to 1/loop, in the case that these clients are slow to process, it will allow more responsiveness to better-behaved clients, and reduce average latency.  These "big command" clients are actually the least latency sensitive because the commands take so long to process, that extra event loop cycles are negligible.

I agree with your point that it's unlikely that multiple will get unblocked at the same time as they are network bound.  So I don't feel too strongly about this.

Reply:
I also don't agree with the one at a time in this case. If the end user is multiplexing a lot of different clients into Redis, it's not strictly true that the overall latency will be better by only allowing 1 at a time. It is true that if we keep deferring clients that produce a lot of output, we could let them accumulate so many more commands that they keep triggering too much output. It seems like a counter intuitive optimization that I would prefer to ignore.

@oranagra Other thought, do you know if there is a compelling reason (other than IO threads) that we don't just write out a clients output buffer after processInputBuffer? It seems like it would lower some of the higher latency seen by deeply pipelined clients.

Reply:
@madolson one of the reasons is the AOF (with or without fsync `always`.
we buffer the propagated commands into a buffer so that we use a single `write` syscall to put them to the AOF (more efficient), and possibly a single `fsync`, but we don't wanna reply to the client before we wrote it to the file (and possibly fsynced it). so if we crash, we know that the client didn't get an ACK on something that wasn't written.
this is also the reason for the `AE_BARRIER` mechanism.

Reply:
@oranagra This issue was bothering me for a very long time actually. Fulfilling this consistency guarantee by putting off the actual write seems to me very fragile and potentially limiting - e.g. if we'll ever want to refactor IO threads and have more independently managed clients.